### PR TITLE
mgr/orchestrator: add `ceph orchestrator wait`

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Fixes http://tracker.ceph.com/issues/39093

* Add `ceph orchestrator wait` that waits for unfinished completions
* Replace `self._orchestrator_wait([completion])` with `yield`

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

The cli then looks like so:

```
$ ceph 'orchestrator' service ls --wait=false
queued.
$ ceph 'orchestrator' osd create localhost:/dev/xxx --wait=false
queued.
$ ceph orchestrator wait                                       
<read op>
    None ceph-mgr localhost None None None
    None ceph-mgr localhost None None None
    None ceph-mgr localhost None None None
    None ceph-mon localhost None None None
    None ceph-mon localhost None None None
    None ceph-mon localhost None None None
    None ceph-osd localhost None None None
    None ceph-osd localhost None None None
    None ceph-osd localhost None None None
    None ceph-osd localhost None None None
create_osds, args=(<orchestrator.DriveGroupSpec object at 0x7f264d4b01d0>, ['localhost']), kwargs={}
    None

```

- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug
- [x] depends on #26404
- [ ] Add list of ref ids to `wait`. I'd like to move that to a different PR, though.

### CODE REVIEW

File: src/pybind/mgr/orchestrator_cli/module.py [hunk 1]
```diff
@@ -161,69 +259,75 @@ def _create_osd(self, svc_arg=None, inbuf=None):
...
         try:
             drive_group.validate(all_hosts)
         except orchestrator.DriveGroupValidationError as e:
-            return HandleCommandResult(-errno.EINVAL, stderr=str(e))
+            yield HandleCommandResult(-errno.EINVAL, stderr=str(e))
+            return
 
-        completion = self.create_osds(drive_group, all_hosts)
-        self._orchestrator_wait([completion])
+        completion = yield self.create_osds(drive_group, all_hosts)
```
Comment (jmolmo) [inline src/pybind/mgr/orchestrator_cli/module.py]: Maybe this should be part of other review/pr, but i think it is worth to comment here:
I think k that if we move : 
- <Args> to <drive group> object stuff
and
- hosts validation
Into each of the backend orchestrators , the code will be more clean  here.  

Look that this method is different from the rest.... here the orchestrator-cli is doing part of the things that the orchcestrators backends should do. If we do not change this... we are forcing to do the same in each client that will use the "osd create" method in no matter what backend orchestrator.

**Comment by sebastian-philipp** [comment]
Who is responsible for making sure, we're not executing things in parallel that should not run in parallel?

File: src/pybind/mgr/orchestrator_cli/module.py [hunk 2]
```diff
@@ -23,48 +42,126 @@ def inner(*args, **kwargs):
...
+        self._completions = []  # type: List[GenComp]
+
     def _select_orchestrator(self):
         return self.get_module_option("orchestrator")
 
+    @CLIWriteCommand('orchestrator wait',
+                     "name=refs,type=CephString,req=false,n=N",
+                     'Waits for all running completions')
+    @handle_exceptions
+    def _wait(self, refs=None):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator_cli/module.py]: This waits synchronously in the event handler thread without a possibility to time-out. Relates to https://tracker.ceph.com/issues/37514

**Comment by sebastian-philipp** [comment]
Fixes: http://tracker.ceph.com/issues/39093

**Comment by sebastian-philipp** [comment]
unstale. I'm not sure if adding this complexity is worth it right now.

**Comment by sebastian-philipp** [comment]
depends on #30262

**Comment by sebastian-philipp** [comment]
State of this PR: Basically finished, but I'm waiting, till there is a need for it as it increases the complexity within orchestrator_cli.py

**Comment by sebastian-philipp** [comment]
unstale

**Comment by liewegas** [comment]
We can probably close this?

**Comment by sebastian-philipp** [comment]
with https://tracker.ceph.com/issues/44205 we finally have a good way to make this monstrosity  obsolete. yay!

**Pull request closed without merging**